File: lib/src/main/java/com/sd/lib/http/RequestManager.java
package com.sd.lib.http;

import android.text.TextUtils;
import android.util.Log;

import com.sd.lib.http.callback.RequestCallback;
import com.sd.lib.http.cookie.ICookieStore;
import com.sd.lib.http.interceptor.IRequestInterceptor;

import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class RequestManager
{
    private static RequestManager sInstance;

    private final Map<RequestTask, RequestInfo> mMapRequest = new ConcurrentHashMap<>();

    private ICookieStore mCookieStore;

    private IRequestInterceptor mRequestInterceptor;
    private IRequestIdentifierProvider mRequestIdentifierProvider;

    private boolean isDebug = false;

    private RequestManager()
    {
    }

    public static RequestManager getInstance()
    {
        if (sInstance == null)
        {
            synchronized (RequestManager.class)
            {
                if (sInstance == null)
                    sInstance = new RequestManager();
            }
        }
        return sInstance;
    }

    public void setDebug(boolean debug)
    {
        isDebug = debug;
        FTaskManager.getInstance().setDebug(debug);
    }

    public boolean isDebug()
    {
        return isDebug;
    }

    /**
     * 设置cookie管理对象
     *
     * @param cookieStore
     */
    public void setCookieStore(ICookieStore cookieStore)
    {
        mCookieStore = cookieStore;
    }

    /**
     * 返回cookie管理对象
     *
     * @return
     */
    public ICookieStore getCookieStore()
    {
        if (mCookieStore == null)
            mCookieStore = ICookieStore.DEFAULT;
        return mCookieStore;
    }

    /**
     * 设置Request标识生成对象
     *
     * @param requestIdentifierProvider
     */
    public void setRequestIdentifierProvider(IRequestIdentifierProvider requestIdentifierProvider)
    {
        mRequestIdentifierProvider = requestIdentifierProvider;
    }

    private IRequestIdentifierProvider getRequestIdentifierProvider()
    {
        if (mRequestIdentifierProvider == null)
        {
            mRequestIdentifierProvider = new IRequestIdentifierProvider()
            {
                @Override
                public String provideRequestIdentifier(IRequest request)
                {
                    return null;
                }
            };
        }
        return mRequestIdentifierProvider;
    }

    //---------- IRequestInterceptor start ----------

    public void setRequestInterceptor(IRequestInterceptor requestInterceptor)
    {
        mRequestInterceptor = requestInterceptor;
    }

    final IRequestInterceptor mInternalRequestInterceptor = new IRequestInterceptor()
    {
        @Override
        public IResponse beforeExecute(IRequest request) throws Exception
        {
            if (mRequestInterceptor != null)
                return mRequestInterceptor.beforeExecute(request);

            return null;
        }

        @Override
        public IResponse afterExecute(IRequest request, IResponse response) throws Exception
        {
            if (mRequestInterceptor != null)
                return mRequestInterceptor.afterExecute(request, response);

            return response;
        }

        @Override
        public void onError(final Exception e)
        {
            if (mRequestInterceptor != null)
            {
                mRequestInterceptor.onError(e);
            } else
            {
                FTask.runOnUiThread(new Runnable()
                {
                    @Override
                    public void run()
                    {
                        throw new RuntimeException(e);
                    }
                });
            }
        }
    };

    //---------- IRequestInterceptor end ----------

    /**
     * 异步执行请求
     *
     * @param request  请求对象
     * @param callback
     */
    public RequestHandler execute(IRequest request, RequestCallback callback)
    {
        return execute(request, false, callback);
    }

    /**
     * 异步执行请求
     *
     * @param request  请求对象
     * @param sequence 是否按顺序一个个执行，true-是
     * @param callback
     * @return
     */
    public synchronized RequestHandler execute(final IRequest request, final boolean sequence, RequestCallback callback)
    {
        if (request == null)
            return null;

        if (callback == null)
        {
            callback = new RequestCallback()
            {
                @Override
                public void onSuccess()
                {
                }
            };
        }

        callback.setRequest(request);
        callback.onPrepare(request);

        final RequestTask task = new RequestTask(request, callback, new RequestTask.Callback()
        {
            @Override
            public void onFinish(RequestTask task)
            {
                removeTask(task);
            }
        });

        final String tag = request.getTag();
        final String requestIdentifier = getRequestIdentifierProvider().provideRequestIdentifier(request);

        final RequestInfo info = new RequestInfo();
        info.tag = tag;
        info.requestIdentifier = requestIdentifier;
        mMapRequest.put(task, info);

        if (isDebug())
        {
            Log.i(RequestManager.class.getName(), "execute"
                    + " task:" + task
                    + " callback:" + callback
                    + " requestIdentifier:" + requestIdentifier
                    + " tag:" + tag
                    + " size:" + mMapRequest.size());
        }

        if (sequence)
        {
            task.submitSequence();
        } else
        {
            task.submit();
        }

        return new RequestHandler(task);
    }

    private synchronized boolean removeTask(RequestTask task)
    {
        if (task == null)
            throw new IllegalArgumentException("task is null");

        final RequestInfo info = mMapRequest.remove(task);
        if (info == null)
            return false;

        if (isDebug())
        {
            Log.i(RequestManager.class.getName(), "removeTask"
                    + " task:" + task
                    + " size:" + mMapRequest.size());
        }

        return true;
    }

    /**
     * 根据tag取消请求
     *
     * @param tag
     * @return 申请取消成功的数量
     */
    public synchronized int cancelTag(final String tag)
    {
        if (tag == null || mMapRequest.isEmpty())
            return 0;

        if (isDebug())
            Log.i(RequestManager.class.getName(), "try cancelTag tag:" + tag);

        int count = 0;
        for (Map.Entry<RequestTask, RequestInfo> item : mMapRequest.entrySet())
        {
            final RequestTask task = item.getKey();
            final RequestInfo info = item.getValue();

            if (tag.equals(info.tag) && task.cancel(true))
                count++;
        }

        if (isDebug())
            Log.i(RequestManager.class.getName(), "try cancelTag tag:" + tag + " count:" + count);

        return count;
    }

    /**
     * 根据Request的唯一标识取消请求{@link IRequestIdentifierProvider}
     *
     * @param request
     * @return
     */
    public synchronized int cancelRequestIdentifier(final IRequest request)
    {
        if (request == null || mMapRequest.isEmpty())
            return 0;

        final String identifier = getRequestIdentifierProvider().provideRequestIdentifier(request);
        if (TextUtils.isEmpty(identifier))
            return 0;

        if (isDebug())
            Log.i(RequestManager.class.getName(), "try cancelRequestIdentifier requestIdentifier:" + identifier);

        int count = 0;
        for (Map.Entry<RequestTask, RequestInfo> item : mMapRequest.entrySet())
        {
            final RequestTask task = item.getKey();
            final RequestInfo info = item.getValue();

            if (identifier.equals(info.requestIdentifier) && task.cancel(true))
                count++;
        }

        if (isDebug())
            Log.i(RequestManager.class.getName(), "try cancelRequestIdentifier requestIdentifier:" + identifier + " count:" + count);

        return count;
    }


    private static final class RequestInfo
    {
        public String tag;
        public String requestIdentifier;
    }
}

File: lib/src/main/java/com/sd/lib/http/RequestManager.kt
package com.sd.lib.http

import android.util.Log
import com.sd.lib.http.callback.RequestCallback
import com.sd.lib.http.cookie.ICookieStore
import com.sd.lib.http.cookie.ModifyMemoryCookieStore
import com.sd.lib.http.interceptor.IRequestInterceptor
import com.sd.lib.http.interceptor.IResultInterceptor
import com.sd.lib.http.parser.DefaultResponseParser
import com.sd.lib.http.parser.IResponseParser
import com.sd.lib.http.utils.HttpUtils
import java.util.concurrent.ConcurrentHashMap

class RequestManager private constructor() {
    private val _mapRequest: MutableMap<RequestTask, RequestInfo> = ConcurrentHashMap()

    /** 调试模式 */
    var isDebug = false

    /** cookie管理对象 */
    var cookieStore: ICookieStore = ModifyMemoryCookieStore()

    /** 请求拦截 */
    var requestInterceptor: IRequestInterceptor? = null

    /** http返回解析器 */
    var responseParser: IResponseParser = DefaultResponseParser()

    /** 结果拦截器 */
    var resultInterceptor: IResultInterceptor? = null

    //---------- IRequestInterceptor start ----------

    internal val internalRequestInterceptor = object : IRequestInterceptor {
        override fun beforeExecute(request: IRequest): IResponse? {
            return try {
                requestInterceptor?.beforeExecute(request)
            } catch (e: Exception) {
                onError(e)
                null
            }
        }

        override fun afterExecute(request: IRequest, response: IResponse): IResponse? {
            return try {
                requestInterceptor?.afterExecute(request, response)
            } catch (e: Exception) {
                onError(e)
                null
            }
        }

        override fun onError(e: Exception) {
            val interceptor = requestInterceptor
            if (interceptor == null) {
                HttpUtils.runOnUiThread { throw RuntimeException(e) }
                return
            }

            try {
                interceptor.onError(e)
            } catch (newE: Exception) {
                HttpUtils.runOnUiThread { throw RuntimeException(newE) }
            }
        }
    }

    //---------- IRequestInterceptor end ----------

    /**
     * 异步执行请求
     *
     * @param sequence 是否按顺序执行
     */
    @JvmOverloads
    @Synchronized
    fun execute(request: IRequest, sequence: Boolean = false, callback: RequestCallback?): RequestHandler {
        val tCallback = callback ?: object : RequestCallback() {
            override fun onSuccess() {}
        }

        // 创建请求任务
        val task = object : RequestTask(request, tCallback) {
            override fun onFinish() {
                removeTask(this)
            }
        }
        val requestHandler = RequestHandler(task)

        tCallback.saveRequest(request)
        tCallback.saveRequestHandler(requestHandler)
        tCallback.onPrepare(request)

        val info = RequestInfo(request.tag)
        _mapRequest[task] = info

        if (isDebug) {
            Log.i(
                RequestManager::class.java.name, "execute"
                        + " task:${task}"
                        + " callback:${tCallback}"
                        + " tag:${info.tag}"
                        + " size:${_mapRequest.size}"
            )
        }

        if (sequence) {
            task.submitSequence()
        } else {
            task.submit()
        }
        return requestHandler
    }

    @Synchronized
    private fun removeTask(task: RequestTask): Boolean {
        _mapRequest.remove(task) ?: return false
        if (isDebug) {
            Log.i(
                RequestManager::class.java.name, "removeTask"
                        + " task:" + task
                        + " size:" + _mapRequest.size
            )
        }
        return true
    }

    /**
     * 根据[tag]取消请求
     *
     * @return 取消的任务数量
     */
    @Synchronized
    fun cancelTag(tag: String?): Int {
        if (tag == null || _mapRequest.isEmpty()) return 0

        if (isDebug) {
            Log.i(RequestManager::class.java.name, "try cancelTag tag:$tag")
        }

        var count = 0
        for ((task, info) in _mapRequest) {
            if (tag == info.tag && task.cancel()) {
                count++
            }
        }

        if (isDebug) {
            Log.i(RequestManager::class.java.name, "try cancelTag tag:$tag count:$count")
        }
        return count
    }

    private class RequestInfo(val tag: String?)

    companion object {
        val instance: RequestManager by lazy { RequestManager() }
    }
}